libtransmission/file-posix.c

```diff
@@ -430,6 +451,132 @@ bool tr_sys_path_rename(char const* src_path, char const* dst_path, tr_error** e
     return ret;
 }
 
+/* We try to do a fast (in-kernel) copy using a variety of non-portable system
+ * calls. If the current implementation does not support in-kernel copying, we
+ * use a user-space fallback instead. */
+bool tr_sys_path_copy(char const* src_path, char const* dst_path, uint64_t file_size, tr_error** error)
+{
+    TR_ASSERT(src_path != NULL);
+    TR_ASSERT(dst_path != NULL);
+
+#if defined(__APPLE__) && defined(HAVE_COPYFILE)
+    (void)file_size; /* unused variable */
+
+    copyfile_flags_t const flags =
+#ifdef COPYFILE_CLONE /* macos >= 10.12 */
+        COPYFILE_CLONE;
+#else
+        COPYFILE_DATA;
+#endif
+    if (copyfile(src_path, dst_path, NULL, flags) < 0)
+    {
+        set_system_error(error, errno);
+        return false;
+    }
+    else
+    {
+        return true;
+    }
+
+#else /* __APPLE__ && HAVE_COPYFILE */
+
+    /* Other OSes require us to copy between file descriptors, so open them. */
+
+    tr_sys_file_t in;
+    tr_sys_file_t out;
+
+    in = tr_sys_file_open(src_path, TR_SYS_FILE_READ |
+        TR_SYS_FILE_SEQUENTIAL, 0, error);
+
+    if (in == TR_BAD_SYS_FILE)
+    {
+        tr_error_prefix(error, "Unable to open old file: ");
+        return false;
+    }
+
+    out = tr_sys_file_open(dst_path, TR_SYS_FILE_WRITE | TR_SYS_FILE_CREATE |
+        TR_SYS_FILE_TRUNCATE, 0666, error);
+
+    if (out == TR_BAD_SYS_FILE)
+    {
+        tr_error_prefix(error, "Unable to open new file: ");
+        tr_sys_file_close(in, NULL);
+        return false;
+    }
+
+#if defined(USE_COPY_FILE_RANGE) || defined(USE_SENDFILE64)
+
+    while (file_size > 0)
+    {
+        size_t const chunk_size = MIN(file_size, SIZE_MAX);
+        ssize_t const copied =
+#ifdef USE_COPY_FILE_RANGE
+            copy_file_range(in, NULL, out, NULL, chunk_size, 0);
+#elif defined(USE_SENDFILE64)
+            sendfile64(out, in, NULL, chunk_size);
+#endif
+
+        if (copied < 0)
+        {
+            set_system_error(error, errno);
+            break;
+        }
+
+        TR_ASSERT(copied >= 0 && ((uint64_t)copied) <= file_size);
+        file_size -= copied;
+    }
+
+#else /* USE_COPY_FILE_RANGE || USE_SENDFILE64 */
+
+    /* Fallback to user-space copy. */
+
+    char* buf = NULL;
+    size_t const buflen = 1024 * 1024; /* 1024 KiB buffer */
+    /* XXX buflen should be configurable. */
+    buf = tr_valloc(buflen);
+    /* XXX handle bad buf */
+
+    while (file_size > 0)
```

Comment:
I couldn't help but notice some similarity between this loop and the one above. Not sure if it makes sense to merge the two, but if there's a way to do this without hurting readability I'd welcome the change.

Another option would be to extract this part into a separate function (accepting either paths or file descriptors) implemented in file.c, given that it's coded in system-independent way based solely on our file wrappers.

Reply:
I looked at this, and I agree in principle, but there's some set up and tear down around the temporary buffer, and the `chunk_size` is different. So the code duplication is either the `while` loop bits (as at present) or the code duplication is the `#if USE_FILE_COPY_RANGE || ...` bits, above, below, and within a single loop to handle those slight differences. With those choices I feel copying the loop infrastructure better expresses the intent of what's going on.